File: w3/w3cpp/Pen_Paper/Pen.cpp
#include <iostream>
#include "Pen.h"
// #include "Paper.h"

Pen::Pen(int inkCapacity) {
    this->inkAmount = inkCapacity;
    this->inkCapacity = inkCapacity;
}

Pen::~Pen() {
    
}

int Pen::getInkAmount() const {
    return this->inkAmount;
}

int Pen::getInkCapacity() const {
    return this->inkCapacity;
}

void Pen::write(Paper& paper, const std::string& message) {
    int inkNeeded = message.size();
    int spaceAmount = paper.getMaxSymbols() - paper.getSymbols();
    
    if ( !this->inkAmount ) {
        throw OutOfInk();
    }
    
    if ( inkNeeded > this->inkAmount ) {
        inkNeeded = this->inkAmount;
    }
    
    if ( inkNeeded > spaceAmount ) {
        inkNeeded = spaceAmount;
    }
    
    this->inkAmount -= inkNeeded;
    
    paper.addContent(message.substr(0, inkNeeded));
}

void Pen::refill() {
    this->inkAmount = this->inkCapacity;
}

File: w3/w3java/Pen_Paper/Pen.java
public class Pen {
    private int inkAmount;
    private int inkCapacity;
    
    public Pen(int inkCapacity) {
        this.inkAmount = inkCapacity;
        this.inkCapacity = inkCapacity;
    }
    
    public Pen() {
        this.inkAmount = 4096;
        this.inkCapacity = 4096;
    }
    
    public int getInkAmount() {
        return this.inkAmount;
    }
    
    public int getInkCapacity() {
        return this.inkCapacity;
    }
    
    public void write(Paper paper, String message) throws OutOfInkException {
        if ( this.inkAmount == 0 ) {
            throw new OutOfInkException();
        }
        
        if ( message.length() > this.inkAmount ) {
            paper.addContent(message.substring(0, this.inkAmount));
            this.inkAmount = 0;
            return;
        }
        
        paper.addContent(message);
        this.inkAmount -= message.length();
    }
    
    public void refill() {
        this.inkAmount = this.inkCapacity;
    }
}